### PR TITLE
Louis/amazon updater

### PR DESCRIPTION
**Pull request by ldelossa**
This PR implements the Amazon Linux ALAS updater. A parser was written in a separate repository https://github.com/quay/alas

### CODE REVIEW

File: aws/client.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+}
+
+// RepoMD returns a alas.RepoMD containing sha256 information of a repositories contents
+func (c *Client) RepoMD() (alas.RepoMD, error) {
+	for _, mirror := range c.mirrors {
+		m := *mirror
+		m.Path = path.Join(m.Path, repoDataPath)
+
+		c.logger.Debug().Msgf("attempting repomd download from mirror %v", m)
+		resp, err := c.c.Get(m.String())
```
Comment (hdonnay) [inline aws/client.go]: this should have a context.Context wired in through here

File: aws/client.go [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+	}
+	err := client.getMirrors(release)
+	return client, err
+}
+
+// RepoMD returns a alas.RepoMD containing sha256 information of a repositories contents
+func (c *Client) RepoMD() (alas.RepoMD, error) {
+	for _, mirror := range c.mirrors {
+		m := *mirror
+		m.Path = path.Join(m.Path, repoDataPath)
```
Comment (hdonnay) [inline aws/client.go]: If this is a url.URL, you can use `.Parse()` to resolve the reference.

File: aws/client.go [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+	return alas.RepoMD{}, fmt.Errorf("all mirrors failed to retrieve repo metadata")
+}
+
+// Updates returns the *http.Response of the first mirror to establish a connection
+func (c *Client) Updates() (*http.Response, error) {
+	for _, mirror := range c.mirrors {
+		m := *mirror
+		m.Path = path.Join(m.Path, updatesPath)
+
+		req, err := http.NewRequest(http.MethodGet, m.String(), nil)
```
Comment (hdonnay) [inline aws/client.go]: should take a context.Context

File: aws/client.go [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+		m := *mirror
+		m.Path = path.Join(m.Path, repoDataPath)
+
+		c.logger.Debug().Msgf("attempting repomd download from mirror %v", m)
+		resp, err := c.c.Get(m.String())
+		if err != nil {
+			c.logger.Error().Msgf("failed to retrieve repomd from mirror %v: %v", m, err)
+			continue
+		}
+		if (resp.StatusCode <= 199) || (resp.StatusCode >= 300) {
```
Comment (hdonnay) [inline aws/client.go]: Why doesn't this just check for `http.StatusOK`? Any other 2xx response would still be weird.

File: aws/client.go [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+	case Linux1:
+		resp, err = c.c.Get(amazonLinux1Mirrors)
+	case Linux2:
+		resp, err = c.c.Get(amazonLinux2Mirrors)
+	}
+
+	if err != nil {
+		c.logger.Error().Msgf("failed to make request for %v mirrors: %v", release, err)
+		return fmt.Errorf("failed to make request for %v mirrors: %v", release, err)
+	}
```
Comment (hdonnay) [inline aws/client.go]: need to close the response Body

File: aws/client.go [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+		c.logger.Info().Msgf("successfully retrieved updates from mirror %v", m)
+		return resp, nil
+	}
+
+	c.logger.Error().Msg("exhausted all mirrors requesting updates")
+	return nil, fmt.Errorf("all update_info mirrors failed to return a response")
+}
+
+func (c *Client) getMirrors(release Release) error {
```
Comment (hdonnay) [inline aws/client.go]: Should take a context.Context

File: aws/client_test.go [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+import (
+	"net/http"
+	"net/url"
+	"testing"
+
+	"github.com/quay/claircore/test/integration"
+	"github.com/stretchr/testify/assert"
+)
+
+func Test_Client_GetMirrors(t *testing.T) {
```
Comment (hdonnay) [inline aws/client_test.go]: The tests that handle the body should have an in-process http server spun up and at least be tested against a known-good body.

File: aws/updater.go [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+	if err != nil {
+		return nil, "", fmt.Errorf("updates repo metadata could not be retrieved: %v", err)
+	}
+
+	resp, err := client.Updates()
+	if err != nil {
+		return nil, "", fmt.Errorf("failed to retrieve update info: %v", err)
+	}
+
+	gzip, err := gzip.NewReader(resp.Body)
```
Comment (hdonnay) [inline aws/updater.go]: This should be buffered into a temporary file.

File: aws/client.go (hunk 2 above)
Comment (ldelossa) [reply aws/client.go]: Going to skip on this one, amazon returns urls which do not end in a slash so the u.Parse() does not to the right thing. https://play.golang.org/p/Uiy1zWJh4sY

I don't want to add a "/" to the end of each string, id prefer just the path.Join method

**Pull request merged**